utils/lstopo/lstopo-tikz.c

```diff
@@ -0,0 +1,154 @@
+{
+  FILE *file = loutput->file;
+  int r = lcolor->r, g = lcolor->g, b = lcolor->b;
+
+  fprintf(file, "\t\\filldraw [fill=hwloc-color-%d-%d-%d,draw=black,line width=1pt] (%u,%u) rectangle ++(%u,%u);\n",
+          r, g, b, x, y, width, height);
+}
+
+
+static void
+tikz_line(struct lstopo_output *loutput, const struct lstopo_color *lcolor, unsigned depth __hwloc_attribute_unused, unsigned x1, unsigned y1, unsigned x2, unsigned y2, hwloc_obj_t obj __hwloc_attribute_unused, unsigned line_id __hwloc_attribute_unused)
+{
+  FILE *file = loutput->file;
+  int r = lcolor->r, g = lcolor->g, b = lcolor->b;
+
+  fprintf(file, "\t\\draw [draw=hwloc-color-%d-%d-%d,line width=1pt] (%u,%u) -- (%u,%u);\n",
+          r, g, b, x1, y1, x2, y2);
+}
+
+static void
+tikz_textsize(struct lstopo_output *loutput __hwloc_attribute_unused, const char *text __hwloc_attribute_unused, unsigned textlength, unsigned fontsize, unsigned *width)
+{
+  fontsize = TIKZ_FONTSIZE_SCALE(fontsize);
+  *width = TIKZ_TEXT_WIDTH(textlength, fontsize);
+}
+
+
+static void
+tikz_text(struct lstopo_output *loutput, const struct lstopo_color *lcolor, int size __hwloc_attribute_unused, unsigned depth __hwloc_attribute_unused, unsigned x, unsigned y, const char *text, hwloc_obj_t obj __hwloc_attribute_unused, unsigned text_id __hwloc_attribute_unused)
+{
+  FILE *file = loutput->file;
+  int r = lcolor->r, g = lcolor->g, b = lcolor->b;
+
+  const char *tikzdelim = "{}%&#";
+
+  fprintf(file, "\t\\node [hwloc-label,text=hwloc-color-%d-%d-%d] at (%u,%u) {",
+          r, g, b, x, y);
+  while (*text) {
+    size_t chunksize = strcspn(text, tikzdelim), n_delim;
+    fprintf(file, "%.*s", (int) chunksize, text);
+    text += chunksize;
+    for (n_delim = strspn(text, tikzdelim); *text && n_delim; ++text, --n_delim)
+      fprintf(file, "\\%c", *text);
+  }
+  fprintf(file, "};\n");
+}
+
+static struct draw_methods tikz_draw_methods = {
+  tikz_declare_color,
+  tikz_box,
+  tikz_line,
+  tikz_text,
+  tikz_textsize,
+};
+
+int output_tikz(struct lstopo_output * loutput, const char *filename)
+{
+  const char *font_family;
+  FILE *output = open_output(filename, loutput->overwrite);
+  if (!output) {
+    fprintf(stderr, "Failed to open %s for writing (%s)\n", filename, strerror(errno));
+    return -1;
+  }
+
+  font_family = tikz_get_font_family();
+  if (!font_family) {
+    /* TODO: choose what to do on error.
+     * For now, silently ignore it
+     */
+    font_family = "";
+  }
+
+  loutput->file = output;
+  loutput->methods = &tikz_draw_methods;
+  loutput->backend_data = output;
+
+  /* recurse once for preparing sizes and positions */
+  loutput->drawing = LSTOPO_DRAWING_PREPARE;
+  output_draw(loutput);
+  loutput->drawing = LSTOPO_DRAWING_DRAW;
+
+  fprintf(output, "\\documentclass{standalone}\n\\usepackage{tikz}\n\\usepackage{xcolor}\n");
+  declare_colors(loutput);
```

Comment:
don't you need something like this before and after declare colors?
%%%% If inserting in another document, the following lines below must be copied before \begin{document} %%%%
%%%% End of lines needed before \begin{document} %%%%%

Reply:
The idea was to have a label easy to parse so the user and to consider he/she would know what to do if he/she decided to edit it manually/extract the image themselves.

But done.